### DISCUSSION TITLE
Scrapli send_commands

### DISCUSSION THREAD

**Discussion by tybruno** [carlmontanari/scrapli#128] [Ideas]
Hello,

Can the send_commands handle all the different configuration modes and interactive mode commands for Cisco IOS devices? I am either going to build a web app or cli that uses send_commands to allow users at run time to input the devices they want to connect to and the commands they want to run. In this case, the tool should be able to run any type of command. Basically, this tool would run a batch of commands on multiple devices.

`send_commands(["show cdp nei", "conf t", "crypto key zeroize","y","exit","write erase","y","show run"]`

Use case:
As a networking engineer, I want to be able to input the cli commands (show , configuration, and interactive) on a list of devices, so that I don't have to login to each device manually to run the same commands.

**Comment by carlmontanari** [comment]
Hey @tybruno 

Sorta kinda this probably could work... on network drivers we compile a pattern that matches all privilege levels of a device, and because of that we shouldn't really ever *not* find a prompt we are happy with. That said, the main idea w/ scrapli things is `send_command(s)` for "command" things and `send_configs` for "config" things... and when you use `send_command(s)` scrapli will auto acquire your platforms `default_desired_privilege_level`, so that kinda throws things outa whack for what you are after...

I think my general recommendation if you are wanting to bypass the "normal" scrapli prompt/privilege level type behavior would be to just create a custom platform based on the `GenericDriver` that has a `comms_prompt_pattern` that matches all privilege levels. Making this would be pretty trivial, and you could even get it added over in scrapli community if you were so inclined. [This](https://github.com/scrapli/scrapli_community/blob/master/scrapli_community/scrapli/genericdriver/scrapli_genericdriver.py) is used just for scrapli testing, but it actually does work w/ a Cisco WLC -- your generic driver setup could pretty much be exactly this, but for the comms prompt pattern you would just have a pattern that matches every privilege on the device.

Hopefully that makes sense! If not, can clarify tomorrow/later after I get some sleep 💤 🙂 

Carl

**Reply by tybruno** [reply]
@carlmontanari  Thank you for the reply!

That is exactly what I am working for. If I have this platform based on this GenericDriver would I have to make it every type of device or is it generic enough that since it doesn't care about the privilege levels it can work with almost all platforms?

Once I have my version of the GenericDriver all I would need to do next is use send_commands with eager = True?

haha thanks again for your help :) I hope you got some good sleep.

**Comment by carlmontanari** [comment]
> If I have this platform based on this GenericDriver would I have to make it every type of device or is it generic enough that since it doesn't care about the privilege levels it can work with almost all platforms?

This really depends on how vague you want to get on the `comms_prompt_pattern`.

The default pattern for the generic driver is `^\S{0,48}[#>$~@:\]]\s*$` -- so its already pretty lose and should match for many devices/privilege levels. 

 Heres a working example for IOSXE (and probably EOS) -- you can see we dont need to do anything to the comms prompt pattern it "just works". 

```
from scrapli.driver import GenericDriver

def main():
    conn = GenericDriver(host="some_host")
    conn.open()
    print(conn.get_prompt())
    conn.send_command("configure terminal")
    print(conn.get_prompt())
    conn.send_command("end")
    print(conn.get_prompt())
    conn.send_command("disable")
    print(conn.get_prompt())

    conn.channel.write("exit")
    conn.channel.send_return()
    conn.close()


if __name__ == "__main__":
    main()

```

You could get crazy with the `comms_prompt_pattern` to make it much more explicit -- or less... whatever you need to do! scrapli's job is basically to work with the pattern you give it and provide you with the api to send whatever ya want! Do note that if the pattern matches too much or too little things can and will break -- so you need to be sure that the pattern you use is solid! Hopefully this will work for ya!

Carl

**Reply by carlmontanari** [reply]
Could have sworn I saw a question about handling secondary auth (enable password) but maybe I'm going nuts.... in the event I did see that:

Probably if you want to take this `GenericDriver` approach to things you are in a bit on your own for handling privilege levels. There is a `send_and_read` method that you could use, you could use `send_interactive` and catch timeouts looking for a password prompt (you'll either need to patch close method -or- just reopen a connection if it times out as it will close the transport in the event of a timeout -- this is intentional to make sure there are no dangling connections), or you could just reach into the channel and use the `read` and `write` methods in there -- `conn.channel.X` and then respond based on the output you see. 

Hopefully that helps! Note that there was at a time a `timeout_exit` flag that I sorta maybe wanna bring back to optionally *not* close the transport on hitting a timeout... may have to think about adding that back :) (or you could obv implement this in a custom setup for yourself if you were so inclined! can find it in git history too im sure)

Carl

**Reply by tybruno** [reply]
Hello Carl, 

Thank you for the response! Sorry for the late reply, I was busy last week and wanted to take a deeper look into Scrapli. 

I would like to keep the ability to log in to the different types of platforms, especially with secondary passwords (i.e. enable). If I did Extend GenericDriver I would have to need to further extend it to handle each platform. This may cause a lot of duplication of code that is already written. 

Here is another possible solution I was thinking of. What if, I either Monkey Patch or MR GenericDriver with another method, let's say cli_shell(commands:list)->MultiResponse, Which would simulate shell interactions on the devices. This way the platforms that inherit from GeneriDriver would have access to this method. 

One thing I think would be important for this method, is to be able to have Response/MultiResponse store the initial_prompt, channel_input, result, and after_prompt. This way after all the CLI commands have been run we can display the results to replicate what would have been seen from the shell. 

I think what Scrapli offers currently is awesome for knowing the commands and states that will be run on the Device before run time. However, this new feature would be unique allowing states and commands to be determined at run time by emulating the platform's cli. I built a script that does this using Netmiko or Scrapli ([code](https://github.com/carlmontanari/scrapli/compare/master...BrunoParadigm:Add-send_any_commands) note this was more of a proof of concept), however, It was a fork, and had to change the source code slightly. I would like to integrate this use case into a real library instead of having a patch. I have created a web app that run's multiple commands on multiple devices using this patch. The networking engineers at Cisco love that tool. Now they would like to see that tool on Cisco DevNet.  I am hoping to use Scrapli as the library that handles the connections and the sending of cli commands. I also hope to promote Scrapli to other developers within Cisco. 


Let me know what you think.

**Reply by carlmontanari** [reply]
Hey @tybruno !

Let me try to recap to make sure I'm tracking what ya wanna do so I don't go off on a weird tangent or something :D 

You ultimately want two things:

1. To be able to send commands I'll say "arbitrarily" -- as in without regard for priv levels -- just pass a list of "stuff" to send and have it send -- something like `conn.send_blah(["conf t", "interface loopback1", "end", "show version", "disable", "enable"])`
2. To have output that looks like exactly what happened in the response object -- like in one of the docstrings in the link you provided.

I think realistically point 1 already works -- I just would not ever really advertise it haha. `send_commands` only checks the privilege level prior to sending the initial command, so something like this works (like I *think/hope* you want for pt1):

```
ipdb> result = conn.send_commands(commands=["show version | i prod_rel", "configure t", "do show run | i hostname", "end", "show version | i reason"])

ipdb> for res in result: print(res.result)
Compiled Tue 18-Sep-18 13:20 by prod_rel_team
Enter configuration commands, one per line.  End with CNTL/Z.
hostname C3560CX

Last reload reason: power-on
```

While I dont know if I think this is a common use case/ask, I would consider adding a `ignore_privelege_level` flag to the `send_command(s|fromfile)` methods of the network driver which I think would make it behave even more like what you are after. Do you see that tweak being useful/valuable?

As far as point 2 goes... thats a tick harder just due to the way scrapli "consumes" the prompt (which is why we dont see the leading prompt in scrapli results and such). *But* you have given me an idea about how to do it as I do think it would be a valuable addition. The idea requires a little back story so bear with me....

Currently you can set the `channel_log` attribute to a string file path, `True` (which just writes to `scrapli_channel.log`) or to a `BytesIO` object. So... basically channel log ends up being a file like object in any of these cases (*if* a user wants to turn it on).  The channel log, since it captures every single read, ends up having a "full" picture of what would have been on the console (hopefully that makes sense).

I *think* (having not mocked it up or anything haha, famous last words) it would be fairly trivial to modify the channel log behavior to basically *always* log to some bytesio object internally, then just write it out to some file if a user requests that to happen. Why am I saying this at all and what does it have to do w/ the response object!? Well... if we wanted, we could call `tell` on the bytesio object at the beginning of a `send_command` method, and then at the end we call tell again, then we could read back the bits between the two positions in the file and bam we have the full output. 

Now that I'm writing this out this sounds like a reasonably solid idea but probably a bit more work than I started off thinking hahah. 

Im going to stop here since thats a lot of brain dump that hopefully makes sense, but will definitely not make sense if I keep going 🤣  -- lemme know your thoughts? 

Thanks for a great conversation (and the scrapli love!)!!

Carl

**Reply by tybruno** [reply]
Hey @carlmontanari!

Thank you for the awesome response!!! :)

1. You hit the nail on the head. That is exactly what I was wanting. I think `ignore_privelege_level` would be useful, although may not be necessary for most use cases, however, it looks very simple to implement. If you would like, I could add that feature and submit an MR? I would like to practice contributing to an open-source project :). 

2. Your backstory makes total sense to me! It seems like it may be fairly challenging, but I also do think it would be a valuable addition for other use cases as well :). I think it is going to be a great addition!!

Points 1 and 2 seem to be great additions. What do you think she be done now?

hahaha thanks again for being so helpful and willing to hear me out. This is an awesome project! I was very excited when I stumbled upon it hahaha. I hope I can contribute to make this an even better project :)

**Reply by carlmontanari** [reply]
> If you would like, I could add that feature and submit an MR? I would like to practice contributing to an open-source project :).

Totally welcome to get PR goin! 

I'll probably have to spend some time this weekend or next weekend looking at pt2... I think we can (and should probably) add the "what you woulda seen if you typed this stuff" output to the response object so probably worth digging in a bit to the channel log/other stuff I mentioned. Ideally I can get some stuff tested out and we can pow wow on it prior to 2021.07.30 since that'll be the next pypi release (doing releases 2x/yr to just keep myself sane), but if we miss that for whatever reason we can ofc just install develop branch (or cut a pre release for 2022.01.30 that you can install with the `--pre` pip option).

So yeah -- PR would be welcome, we can pow wow on that and get that sorted (tests please!) and then over next few weeks I'll look at the other bits and we'll see what shakes out!

**Reply by tybruno** [reply]
Awesome :) Sounds perfect! Thank you! I submitted an #137 issue for the PR. I'll chat with you at the pow wow :)

**Reply by tybruno** [reply]
Hello @carlmontanari,

I saw that the new version of Scrapli is out now!! Yaay 🎆 !! In this new version were you able to add the ability for the response objects to replicate the cli prompt and output?  See point 2 for the original conversation. I know it's been a long time ahahah 🤣 

> Hey @tybruno !
> 
> Let me try to recap to make sure I'm tracking what ya wanna do so I don't go off on a weird tangent or something :D
> 
> You ultimately want two things:
> 
> 1. To be able to send commands I'll say "arbitrarily" -- as in without regard for priv levels -- just pass a list of "stuff" to send and have it send -- something like `conn.send_blah(["conf t", "interface loopback1", "end", "show version", "disable", "enable"])`
> 2. To have output that looks like exactly what happened in the response object -- like in one of the docstrings in the link you provided.
> 
> I think realistically point 1 already works -- I just would not ever really advertise it haha. `send_commands` only checks the privilege level prior to sending the initial command, so something like this works (like I _think/hope_ you want for pt1):
> 
> ```
> ipdb> result = conn.send_commands(commands=["show version | i prod_rel", "configure t", "do show run | i hostname", "end", "show version | i reason"])
> 
> ipdb> for res in result: print(res.result)
> Compiled Tue 18-Sep-18 13:20 by prod_rel_team
> Enter configuration commands, one per line.  End with CNTL/Z.
> hostname C3560CX
> 
> Last reload reason: power-on
> ```
> 
> While I dont know if I think this is a common use case/ask, I would consider adding a `ignore_privelege_level` flag to the `send_command(s|fromfile)` methods of the network driver which I think would make it behave even more like what you are after. Do you see that tweak being useful/valuable?
> 
> As far as point 2 goes... thats a tick harder just due to the way scrapli "consumes" the prompt (which is why we dont see the leading prompt in scrapli results and such). _But_ you have given me an idea about how to do it as I do think it would be a valuable addition. The idea requires a little back story so bear with me....
> 
> Currently you can set the `channel_log` attribute to a string file path, `True` (which just writes to `scrapli_channel.log`) or to a `BytesIO` object. So... basically channel log ends up being a file like object in any of these cases (_if_ a user wants to turn it on). The channel log, since it captures every single read, ends up having a "full" picture of what would have been on the console (hopefully that makes sense).
> 
> I _think_ (having not mocked it up or anything haha, famous last words) it would be fairly trivial to modify the channel log behavior to basically _always_ log to some bytesio object internally, then just write it out to some file if a user requests that to happen. Why am I saying this at all and what does it have to do w/ the response object!? Well... if we wanted, we could call `tell` on the bytesio object at the beginning of a `send_command` method, and then at the end we call tell again, then we could read back the bits between the two positions in the file and bam we have the full output.
> 
> Now that I'm writing this out this sounds like a reasonably solid idea but probably a bit more work than I started off thinking hahah.
> 
> Im going to stop here since thats a lot of brain dump that hopefully makes sense, but will definitely not make sense if I keep going 🤣 -- lemme know your thoughts?
> 
> Thanks for a great conversation (and the scrapli love!)!!
> 
> Carl

**Reply by carlmontanari** [reply]
Hey @tybruno !

> In this new version were you able to add the ability for the response objects to replicate the cli prompt and output?

No unfortunately. I've been pretty busy and that'll probably be a fairly big effort with potentially fairly big impact (scrapli-netconf, scrapligo and of course "normal" scrapli things). Def still on my radar though!

**Reply by tybruno** [reply]
@carlmontanari No worries, I noticed you made a lot of good changes this version. If you do support that feature please let me know. I would definitely use it and would help test it.

**Reply by carlmontanari** [reply]
> @carlmontanari No worries, I noticed you made a lot of good changes this version. If you do support that feature please let me know. I would definitely use it and would help test it.

Will do dude!

**Reply by tybruno** [reply]
@carlmontanari  hey man! Hope you are doing well. Just wondering, is this feature still on your radar?

**Reply by carlmontanari** [reply]
hey dude! probably not to be honest. im not doing anything related to networking day-to-day anymore, defo would help on getting pr merged but to be honest it would probably be a fairly big one so that may be rough. I think shorter term/best option is to send channel log to a buffer and slurp what you want out of that!

**Reply by tybruno** [reply]
No worries :) Thank you for the quick response. Sounds good. I'll look into that.
